core/src/main/scala/kafka/tools/ConsoleProducer.scala

```diff
@@ -110,6 +110,8 @@ object ConsoleProducer {
       props, ProducerConfig.SEND_BUFFER_CONFIG, config.options, config.socketBufferSizeOpt)
     CommandLineUtils.maybeMergeOptions(
       props, ProducerConfig.BUFFER_MEMORY_CONFIG, config.options, config.maxMemoryBytesOpt)
+    CommandLineUtils.maybeMergeOptions(
+      props, ProducerConfig.BATCH_SIZE_CONFIG, config.options, config.maxPartitionMemoryBytesOpt)
```

Comment:
We can try to merge the `config.batchSizeOpt` first, and then merge `config.maxPartitionMemoryBytesOpt`. Like this:

```java
    CommandLineUtils.maybeMergeOptions(
      props, ProducerConfig.BATCH_SIZE_CONFIG, config.options, config.batchSizeOpt)
    CommandLineUtils.maybeMergeOptions(
      props, ProducerConfig.BATCH_SIZE_CONFIG, config.options, config.maxPartitionMemoryBytesOpt)
```

This way, the batch.size value will be replaced with `maxPartitionMemoryBytesOpt` option value (if set), otherwise, the `batchSizeOpt` option value will be used (if set), otherwise, default value will be used (i.e. 16 * 1024). 

These following cases will have such results:
1. bin/kafka-console-producer.sh --bootstrap-server localhost:9092 --topic test **--batch-size 123**
--> batch.size = 123
2. bin/kafka-console-producer.sh --bootstrap-server localhost:9092 --topic test **--max-partition-memory-bytes 456**
--> batch.size = 456
3. bin/kafka-console-producer.sh --bootstrap-server localhost:9092 --topic test **--max-partition-memory-bytes 456 --batch-size 123**
--> batch.size = 456
4. bin/kafka-console-producer.sh --bootstrap-server localhost:9092 --topic test 
--> batch.size = 16 * 1024

Does that make sense?
If so, please help add some tests in `ConsoleProducerTest` for the above 4 cases. 
Thank you.